runtime/near-evm-runner/src/utils.rs

```diff
@@ -200,6 +205,137 @@ pub fn method_name_to_rlp(method_name: String) -> [u8; 4] {
     result
 }
 
+pub fn encode_address(addr: Address) -> Vec<u8> {
+    let mut bytes = vec![0u8; 12];
+    bytes.extend_from_slice(&addr.0);
+    bytes
+}
+
+pub fn encode_string(s: &str) -> Vec<u8> {
+    let mut bytes = vec![];
+    bytes.extend_from_slice(&keccak(s.as_bytes()).as_bytes());
+    bytes
+}
+
+struct Arg {
+    #[allow(dead_code)]
+    name: String,
+    t: String,
+}
+
+struct Method {
+    name: String,
+    args: Vec<Arg>,
+}
+
+impl Arg {
+    fn parse_arg(text: &str) -> Result<(Arg, &str)> {
+        let (t, remains) = parse_ident(text)?;
+        let remains = consume(remains, ' ')?;
+        let (name, remains) = parse_ident(remains)?;
+        Ok((Arg { t, name }, remains))
+    }
+}
+
+impl Method {
+    fn parse_method_args(text: &str) -> Result<(Vec<Arg>, &str)> {
```

Comment:
Let's write some tests for these methods directly.
Something where you have few levels of indirection like `Blah(Foo q, Bar w)Foo(uint256 x, Bar z)Bar(T ttt)T(string x)`

Reply:
done, as eip-712 spec, it require format to no trailing space after `,` and eip-712 digest such no trailing space method name format. If this arg is from metamask, then the input is already no trailing space after `,`. If it's from user, code here raise a parsing error

Reply:
more corner case such as only one method, method without no indirection, method with one indirection is implicitly test by test_meta_call_sig_and_recover. Also the naming "method_name" and "Method" is quite confusing here. "method_name" means both method name and things like `Blah(Foo q, Bar w)Foo(uint256 x, Bar z)Bar(T ttt)T(string x)`. `struct Method` actually means `MethodOrType` I guess, but `Vec<Method>` is sufficient to represent the parsed structure.